e3nn_jax/_src/s2grid.py

```diff
@@ -187,45 +195,58 @@ def from_s2grid(x: jnp.ndarray, lmax: int, normalization="component", lmax_in=No
     else:
         raise Exception("normalization needs to be 'norm', 'component' or 'integral'")
 
-    m = _expand_matrix(range(lmax + 1))  # [l, m, i]
-    shb = _rollout_sh(shb, lmax)
+    # get quadrature weights
+    if quadrature == "soft":
+        assert res_beta % 2 == 0, "res_beta needs to be even for soft quadrature weights to be computed properly"
+        qw = _quadrature_weights_soft(res_beta // 2) * res_beta**2  # [b]
+    elif quadrature == "gausslegendre":
+        _, qw = np.polynomial.legendre.leggauss(res_beta)
+        qw /= 2
+    else:
+        raise Exception("quadrature needs to be 'soft' or 'gausslegendre'")
 
-    assert res_beta % 2 == 0, "res_beta needs to be even for quadrature weights to be computed properly"
-    qw = _quadrature_weights(res_beta // 2) * res_beta**2 / res_alpha  # [b]
-    # beta integrand
-    shb = jnp.einsum("lmj,bj,lmi,l,b->mbi", m, shb, m, n, qw)  # [m, b, i]
+    # prepare beta integrand
+    m = _expand_matrix(range(lmax + 1))  # [l, m, i]
+    shz = _rollout_sh(shz, lmax)
+    shz = jnp.einsum("lmj,bj,lmi,l,b->mbi", m, shz, m, n, qw)  # [m, b, i]
 
     size = x.shape[:-2]
     x = x.reshape(-1, res_beta, res_alpha)
 
     # integrate over alpha
-    int_a = rfft(x, lmax)  # [..., res_beta, 2*l+1]
+    int_a = rfft(x, lmax) / res_alpha  # [..., res_beta, 2*l+1]
     # integrate over beta
-    int_b = jnp.einsum("mbi,zbm->zi", shb, int_a)
-    return int_b.reshape(*size, int_b.shape[1])
+    int_b = jnp.einsum("mbi,zbm->zi", shz, int_a)
+
+    # convert to IrrepsArray
+    coeffs = int_b.reshape(*size, int_b.shape[1])
+    irreps = [(1, (l, p_val * p_arg**l)) for l in range(lmax + 1)]
+    return IrrepsArray(irreps, coeffs)
 
 
-def to_s2grid(coeffs: jnp.ndarray, res=None, normalization="component"):
+def to_s2grid(tensor: e3nn_jax.IrrepsArray, res=None, normalization="component", *, quadrature: str):
     r"""Transform spherical tensor into signal on the sphere
 
     The inverse transformation of :func:`e3nn_jax.from_s2grid`
 
     Args:
-        coeffs (`jax.numpy.ndarray`): array of coefficients, of shape ``(..., (lmax+1)^2)``
+        tensor (`e3nn_jax.IrrepsArray`): spherical tensor, with coefficient array of shape ``(..., (lmax+1)^2)``
         res (tuple, optional): resolution of the grid on the sphere ``(beta, alpha)``
         normalization ({'norm', 'component', 'integral'}): normalization of the spherical tensor
+        quadrature (str): "soft" or "gausslegendre"
 
     Returns:
         `jax.numpy.ndarray`: signal on the sphere of shape ``(..., beta, alpha)``
     """
+    coeffs = tensor.array
```

Comment:
Please add a check for the `tensor.irreps` and raise helpful error for the user. Check that l go from 0 to some lmax and that the parity is one of the valid ones